### DISCUSSION TITLE
Add 2FA support

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#156] [Anything Else]
"As a company that gets many customers because they value the privacy and data security of their customers and themselves, it would be a great addition to Plausible if 2-factor-authentication was available as well as multiple user accounts. Sharing one user account across different employees is by default insecure and needless to say 2FA provides higher security."

**Update:** You can now enable two-factor authentication (2FA) as an extra security layer for your Plausible account in your account settings. Here’s how: https://plausible.io/docs/2fa

**Comment by metmarkosaric** [comment]
Another vote for 2FA via email:

"I'm sure more companies would be attracted by your service if you could offer it. It should be pretty easy and " no cost" (besides implementation/maintain of code of course.. which we all know is a cost :)) to implement a simple solution with TOTP (no need for you to pay per transaction for SMS-codes etc). We did it ourselves in less than a day for our own web service. However I understand it's a question of prioritization. The problem is probably that many doesn't take the time to contact you to check if you have 2FA (as the default guess is "no") so you are probably missing out some business opportunities as well.. Anyway, I would be glad to use your service once you have support for it, so I will check back on the thread."

**Reply by metmarkosaric** [reply]
Another vote for 2FA today which seems to be a popular feature request from companies

**Comment by finnsnape** [comment]
Any update on this? Seems like a critical feature

**Reply by metmarkosaric** [reply]
No update on this at the moment. We'd like to build this but I cannot make any promise on when we will be able to work on it. We're a small team with limited development resources so need to take it one step at a time. There are some other more frequently requested features that we are working on at the moment. Thanks for understanding!

**Reply by alexanderadam** [reply]
I know that [issue 169](https://github.com/plausible/analytics/issues/169) was closed because you have "_have no intention to add any third-party logins_" but maybe it would still be an elegant way also to support external logins and [feature 383](https://github.com/plausible/analytics/discussions/383) might solve this as well anyway, right?

Sign ins for the well known I18n Elixir app [Accent via external providers can be setup via few environment variables](https://github.com/mirego/accent#authentication-setup) and since they're using [Überauth](https://github.com/ueberauth/ueberauth#%C3%BCberauth) adding [another provider strategy like GitLab](https://github.com/mirego/accent/pull/264/files#diff-e43c5cbf91db7a8062b6cb860cbf118296c1b4c7ee32fdcf702e54234ba38092) seem to be straightforward.

Not quite relevant but still related: I saw that Plausible already integrates [OAuther](https://github.com/lexmag/oauther#usage) (for OAuth 1.0) as a [mix dependency](https://github.com/plausible/analytics/blob/948de2b4c9850d7dd990d6f6b456496b99cec9ac/mix.exs#L82) but I'm not able to see where it might be used?

Also thank you so much for working on Plausible. It's highly appreciated and of course it is absolutely understandable if features take time.

**Reply by finnsnape** [reply]
> No update on this at the moment. We'd like to build this but I cannot make any promise on when we will be able to work on it. We're a small team with limited development resources so need to take it one step at a time. There are some other more frequently requested features that we are working on at the moment. Thanks for understanding!

That's a shame, though I appreciate people have different priorities. I love everything else about Plausible and was just about to sign up, but as a freelancer I'm worried about exposing the data of dozens of my clients or leaving it vulnerable to being deleted by someone malicious. You're doing great work regardless though.

In my view, login via email and password is not very viable in `{{currentYear}}`. Is there any plans to add some trivial security measures such as an IP whitelist for admin accounts? Or perhaps I missed some of these features?

**Reply by finnsnape** [reply]
Also with regards to this perhaps not being a frequently requested feature, I think the person you quote above is correct in saying

> The problem is probably that many doesn't take the time to contact you to check if you have 2FA (as the default guess is "no") so you are probably missing out some business opportunities as well

Seems like there may be a silent majority who simply move on when they see this feature does not exist

**Reply by metmarkosaric** [reply]
We don't have any plans to work on this at the moment as we have some other priorities but it is on our list of things to do. We'll update this post when there's something new on this topic.

**Comment by finnsnape** [comment]
Once this is being worked on, I think it would be very helpful to include a feature that enables admins to force 2FA on users they invite

**Reply by metmarkosaric** [reply]
that could be nice!

**Comment by latvia234** [comment]
Please add support for YubiKey and other U2F Security Keys to Plausible. Security keys are the most secure form of 2-Step Verification and protect against phishing threats.

Some interesting articles:

- Wiki: https://en.wikipedia.org/wiki/Universal_2nd_Factor
- Companies that already support Yubikeys: https://www.yubico.com/nl/works-with-yubikey/catalog/?sort=popular
- Interesting article: https://www.yubico.com/resources/reference-customers/google/
- Interesting support article about security keys: https://support.google.com/accounts/answer/6103523

**Comment by MacStasy** [comment]
Any update on this?

**Comment by susanthenerd** [comment]
As I really need 2FA, I plan on implementing it myself in the codebase.
Please let me know if you have any idea on how it should be done.

**Reply by ruslandoga** [reply]
@surelyreformer https://github.com/dashbitco/nimble_totp might be useful :)

**Reply by susanthenerd** [reply]
@ruslandoga Thanks for the info. When adding it, should I add it just for login as an option, or it should ask and for other critical options(change of important account options as an example) for a 2FA code?

**Reply by ruslandoga** [reply]
I think you'd need to present the seed on registration for user to save it (to 1password, for example), and then as an additional check on login. As for asking for it on other actions, I don't know. Does Plausible ask for password anywhere outside of the login page?

**Reply by susanthenerd** [reply]
So you wish to force users to use 2FA? I find it as a possible inconvenience. 

What about backup codes? Anything that should I take in consideration?

**Reply by ruslandoga** [reply]
> So you wish to force users to use 2FA?

That'd be up to you :)

> What about backup codes?

That'd also be up to you, backup codes are not strictly part of totp. Access restoration can happen via backup codes / email / support. Backup codes can be stored like passwords are stored right now.

**Reply by susanthenerd** [reply]
Ok, I will implement TOTP for now and then if others ask about backup codes, I will just add support for them.

**Comment by metmarkosaric** [comment]
**Update:** You can now enable two-factor authentication (2FA) as an extra security layer for your Plausible account in your account settings. Here’s how: https://plausible.io/docs/2fa

**Reply by KaKi87** [reply]
Hi,
I just installed a self-hosted Plausible instance and am not seeing this feature.
Did I miss an env variable or something ?
Thanks

**Reply by ruslandoga** [reply]
👋 @KaKi87 

TOTP was added after v2.0.0 release. It will be present in the upcoming 2.1.0 release though.
